crates/net/network/src/peers/manager.rs

```diff
@@ -262,12 +262,24 @@ impl PeersManager {
 
     /// Removes the tracked node from the set.
     pub(crate) fn remove_discovered_node(&mut self, peer_id: PeerId) {
-        if let Some(entry) = self.peers.remove(&peer_id) {
+        if let Some(mut peer) = self.peers.remove(&peer_id) {
             trace!(target : "net::peers",  ?peer_id, "remove discovered node");
-            if entry.state.is_connected() {
-                // TODO(mattsse): is this right to disconnect peers?
-                self.connection_info.decr_state(entry.state);
-                self.queued_actions.push_back(PeerAction::Disconnect { peer_id, reason: None })
+
+            if peer.state.is_connected() {
+                debug!(target : "net::peers",  ?peer_id, "disconnecting on remove from discovery");
+                // we terminate the session here, so we don't need to keep track of an additional
+                // state where the peer is connected but not in the set. Otherwise, there could be a
+                // scenario in which the peer is discovered again, added to the set again and ready
+                // to make an outgoing connection, even though we might still be connected.
+                // We only mark the peer as disconnecting and add it back to prevent an edge case
+                // where we're disconnecting but the peer is discovered again
+                // immediately.
+                peer.state.disconnect();
+                self.peers.insert(peer_id, peer);
```

Comment:
Happy path: remove, it's not connected, function ends

This specific "insert" acts as a "check" to prevent a race-condition where we asked [..]

Let's add a couple notes  or a unit test for this

Reply:
added unit test, docs and additional check.